polygeist/tools/cgeist/Lib/CGStmt.cc

```diff
@@ -1138,10 +1138,10 @@ ValueCategory MLIRScanner::VisitReturnStmt(clang::ReturnStmt *stmt) {
 
       auto postTy = returnVal.getType().cast<MemRefType>().getElementType();
       if (auto prevTy = val.getType().dyn_cast<mlir::IntegerType>()) {
-        auto ipostTy = postTy.cast<mlir::IntegerType>();
-        if (prevTy != ipostTy) {
-          val = builder.create<arith::TruncIOp>(loc, ipostTy, val);
-        }
+        const auto SrcTy = stmt->getRetValue()->getType();
+        const auto IsSigned =
+            SrcTy->isBooleanType() ? false : SrcTy->isSignedIntegerType();
+        val = rv.IntCast(builder, postTy, IsSigned).val;
```

Comment:
This code was **always** trying to truncate when integer types differed, e.g., `i1 -> i8` in our tests. Now we use the newly introduced `IntCast`.